invokeai/app/invocations/prompt_tools.py

```diff
@@ -0,0 +1,286 @@
+# 2023 skunkworxdark (https://github.com/skunkworxdark)
+
+from typing import Literal, Union, Optional
+import re
+import json
+
+from pydantic import BaseModel
+
+from invokeai.app.invocations.primitives import StringOutput
+from invokeai.app.invocations.baseinvocation import (
+    BaseInvocation,
+    BaseInvocationOutput,
+    InputField,
+    Input,
+    OutputField,
+    InvocationContext,
+    UIComponent,
+    UIType,
+    invocation,
+    invocation_output,
+)
+
+
+@invocation_output("prompt_to_file_output")
+class PromptsToFileInvocationOutput(BaseInvocationOutput):
+    """Base class for invocation that writes to a file and returns nothing of use"""
+
+
+@invocation("prompt_to_file", title="Prompts To File", tags=["prompt", "file"], category="prompt", version="1.0.0")
+class PromptsToFileInvocation(BaseInvocation):
+    """Save prompts to a text file"""
+
+    file_path: str = InputField(description="Path to prompt text file")
+    prompts: Union[str, list[str], None] = InputField(
+        default=None, description="Prompt or collection of prompts to write", ui_type=UIType.Collection
+    )
+    append: bool = InputField(default=True, description="Append or overwrite file")
+
+    def invoke(self, context: InvocationContext) -> PromptsToFileInvocationOutput:
+        with open(self.file_path, "a" if self.append else "w") as f:
+            if isinstance(self.prompts, list):
+                for line in self.prompts:
+                    f.write(line + "\n")
+            else:
+                f.write((self.prompts or "") + "\n")
+
+        return PromptsToFileInvocationOutput()
+
+
+@invocation_output("prompt_pos_neg_output")
+class PromptPosNegOutput(BaseInvocationOutput):
+    """Base class for invocations that output a posirtive and negative prompt"""
+
+    positive_prompt: str = OutputField(description="Positive prompt")
+    negative_prompt: str = OutputField(description="Negative prompt")
+
+
+@invocation(
+    "prompt_split_neg",
+    title="Prompt Split Negative",
+    tags=["prompt", "split", "negative"],
+    category="prompt",
+    version="1.0.0",
+)
+class PromptSplitNegInvocation(BaseInvocation):
+    """Splits prompt into two prompts, inside [] goes into negative prompt everthing else goes into positive prompt. Each [ and ] character is replaced with a space"""
```

Comment:
I wonder if this could be better as a generic string operation that splits a string on a specific string. 

In that case, the prompt could be like "my positive prompt stuff || my negative prompt stuff" and you'd tell the node to split on "||".

I understand this goes against the convention of nesting the negative prompt within `[...]` but parsing nested stuff is never fun and often has edge cases. I can't understand the logic here without stepping through it line by line - the intent isn't quite clear.

A simple split is much easier to reason about than the logic here, which I'm sure works, but which I cannot really wrap my head around without stepping through it line by line.

Anyways, just thinking out loud

Reply:
I think a generic string split would be a great new node and would have many use cases and should be part of a new strings.py 

However this node came in as a bridging idea between the old Invoke and new invoke as I had many old style prompts and wanted to test them against different models and parameters.  I agree that nesting can and does have many edge cases and I apologise if my code is not as clear in intent as is could be but it did grow in an organic manner as users of it asked for additional bits for it. Originally it was almost just a simple split but people wanted to do thing like combine multiple prompts and still have the neg side of it pulled out. So it just grew in complexity.

Reply:
Just so I understand what is going on here.  Did you want me to make these name changes and move them into a new strings.py file?    What would you like to do with the PromptsToFileInvocation?  I am thinking that this should also become StringsToFile and the existing core version (prompts.py) of PromptsFromFileInvocation should also move and be renamed.   What's your thoughts?

Reply:
New node StringSplit added that will split a string into two strings based upon the first occurance of the delimiter
Changed the name of the PromptSplitNegative to StringsplitNegative and moved into strings.py